Donor Profile authentication logic is now implemented

henryholtgeerts opened the pull request:
Resolves #5567

## Description
This PR implements backend and frontend logic to support authentication in the Donor Profile application. Previously, the app would break when viewed by donors who were not logged in. Now, when not logged in, donors are prompted to log in or enter an email address to create an account.

## Affects
This PR affects the REST routes available for the Donor Profile domain (adding Login, Logout) and the frontend rendering logic, requiring that donors log in to view their profile.

## Visuals
<img width="970" alt="Screen Shot 2021-01-26 at 2 23 41 PM" src="https://user-images.githubusercontent.com/5186078/105894680-885d5e00-5fe2-11eb-9bc2-2b4b8aa28f32.png">

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions
1. Add a new page with the Donor Profile block
2. View the page from the frontend, logged in. Check that Profile functions as expected (can modify profile, view donation history, etc).
3. Logout, and view the page. Check that you are prompted to login, and that login fields work as expected.

henryholtgeerts commented:
@ravinderk So far, I have succesfully implemented the basic auth components here (logging in with login and password). I'm now moving on to implementing logic to support the other auth functionality (sending email to access donor profile). I have an idea of how to put this in place, but I'm curious if GiveWP already supports this, before I head off to build out my own implementation.

Do you have any insights?

ravinderk commented:
@henryholtgeerts You can view [`Give_Email_Access`](https://github.com/impress-org/givewp/blob/9cf42048f970424f7ca4e208832cc7bcd7f3448d/includes/class-give-email-access.php#L2) class to understand how we authorize donor and send access email.
The primary use of this is to allow donors to view receipts. Apart from this, I can see recurring donations also utilize this to allow the donor to see the subscription list and authorize them to perform subscription actions.
GiveWP authorizes donor by cookie and db session with help of tokenized URL which donor gets in an email when request access.

henryholtgeerts commented:
@ravinderk Thanks for the help, really great explanation and starting point for me! 👍

ravinderk commented on src/DonorProfiles/Routes/VerifyEmailRoute.php:
```diff
@@ -0,0 +1,110 @@
+
+			if ( $sent === true ) {
+
+				$sentMessage = (string) apply_filters( 'give_email_access_mail_send_notice', __( 'Please check your email and click on the link to access your complete donation history.', 'give' ) );
```
@henryholtgeerts use `esc_html__`

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/46970f78793232cb9f6a34ba3f63077dce49502c

ravinderk commented on src/DonorProfiles/Routes/VerifyEmailRoute.php:
```diff
@@ -0,0 +1,110 @@
+
+			} else {
+
+				$failedMessage = __( 'Unable to send email. Please try again.', 'give' );
```
use `esc_html__`

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/46970f78793232cb9f6a34ba3f63077dce49502c

ravinderk commented on src/DonorProfiles/Routes/VerifyEmailRoute.php:
```diff
@@ -0,0 +1,110 @@
+			$spamMessage = (string) apply_filters(
+				'give_email_access_requests_exceed_notice',
+				sprintf(
+					__( 'Too many access email requests detected. Please wait %s before requesting a new donation history access link.', 'give' ),
```
use `esc_html__`

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/46970f78793232cb9f6a34ba3f63077dce49502c

ravinderk reviewed (changes requested):
1. User login working fine for me but email verification gives me the same error for every donor email:
![image](https://user-images.githubusercontent.com/1784821/107028551-b8bbae00-67d3-11eb-837c-6c6105de98e2.png)
2. Email verification option must be available on if `Email Access` enabled:
![image](https://user-images.githubusercontent.com/1784821/107028839-1c45db80-67d4-11eb-9958-a8eb21c05130.png)
3. Do we show error code or error message:
![image](https://user-images.githubusercontent.com/1784821/107029006-58793c00-67d4-11eb-9dbb-a48d1723482c.png)

henryholtgeerts commented:
@ravinderk Thanks for the review! I've addressed the suggested use of `esc_html__`. I'm working now on addressing the issues you pointed out with the screenshots - very helpful! I'll ping you here and in Slack when the PR is ready for another reiview 👍

henryholtgeerts commented:
@ravinderk All the additional points you raised have been addressed. See the commits related to each:

1. https://github.com/impress-org/givewp/pull/5569/commits/81effebd574c2fc551f9465146756fadd6473535
2. https://github.com/impress-org/givewp/pull/5569/commits/0abe8bbe2841fdacfdc4b5bb7e21373957cf8336, https://github.com/impress-org/givewp/pull/5569/commits/ee1b3066defc18dde3b1ec8d7670e7d72677c75c, https://github.com/impress-org/givewp/pull/5569/commits/950962dda961a230a6c91c0929c83509d092c32f
3. https://github.com/impress-org/givewp/pull/5569/commits/c98f2afd1eacd8f3b633f63b3722ef185b93533d

ravinderk commented on src/DonorProfiles/Exceptions/DuplicateTabException.php:
```diff
@@ -0,0 +1,24 @@
+use Exception;
+
+/**
+ * Define a custom exception class
```
@henryholtgeerts `@since` tag is missing from most of the classes. Can you add?

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/4f6b431d0fead61d30cbae701f7c0ce7b65b94a5

ravinderk commented on src/DonorProfiles/Routes/LogInRoute.php:
```diff
@@ -0,0 +1,102 @@
+						'status'        => 400,
+						'response'      => 'incorrect_password',
+						'body_response' => [
+							'message' => 'The provided password was incorrect.',
```
@henryholtgeerts is this user-facing string. Can we localize it?

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/9950b0fb611831626905aa45c4b258956d94e0ab

ravinderk commented on src/DonorProfiles/Routes/LogInRoute.php:
```diff
@@ -0,0 +1,102 @@
+					'status'        => 400,
+					'response'      => 'unidentified_login',
+					'body_response' => [
+						'message' => "A record for the provided login ('{$login}') could not be found.",
```
@henryholtgeerts is this user-facing string. Can we localize it?

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/9950b0fb611831626905aa45c4b258956d94e0ab

ravinderk commented:
@henryholtgeerts and @ravinderk had discussion about integrating email access logic for donor. @henryholtgeerts will implement it.

ravinderk commented:
@henryholtgeerts I am getting following error when open any admin page:
```
[16-Feb-2021 15:49:39 UTC] PHP Notice:  register_rest_route was called <strong>incorrectly</strong>. The REST API route definition for <code>give-api/v2/donor-profile/login</code> is missing the required <code>permission_callback</code> argument. For REST API routes that are intended to be public, use <code>__return_true</code> as the permission callback. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 5.5.0.) in /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php on line 5311
[16-Feb-2021 15:49:39 UTC] PHP Stack trace:
[16-Feb-2021 15:49:39 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/wp-admin/post-new.php:0
[16-Feb-2021 15:49:39 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/wp-admin/post-new.php:72
[16-Feb-2021 15:49:39 UTC] PHP   3. array_reduce() /Users/ravinderkumar/Local-Sites/give/app/public/wp-admin/edit-form-blocks.php:82
[16-Feb-2021 15:49:39 UTC] PHP   4. rest_preload_api_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-admin/edit-form-blocks.php:82
[16-Feb-2021 15:49:39 UTC] PHP   5. rest_do_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:2520
[16-Feb-2021 15:49:39 UTC] PHP   6. rest_get_server() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:479
[16-Feb-2021 15:49:39 UTC] PHP   7. do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:521
[16-Feb-2021 15:49:39 UTC] PHP   8. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:484
[16-Feb-2021 15:49:39 UTC] PHP   9. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[16-Feb-2021 15:49:39 UTC] PHP  10. Give\Helpers\Hooks::Give\Helpers\{closure:/Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Helpers/Hooks.php:29-38}() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[16-Feb-2021 15:49:39 UTC] PHP  11. call_user_func_array:{/Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Helpers/Hooks.php:37}() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Helpers/Hooks.php:37
[16-Feb-2021 15:49:39 UTC] PHP  12. Give\DonorProfiles\Routes\LoginRoute->registerRoute() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/Helpers/Hooks.php:37
[16-Feb-2021 15:49:39 UTC] PHP  13. register_rest_route() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Routes/LoginRoute.php:27
[16-Feb-2021 15:49:39 UTC] PHP  14. _doing_it_wrong() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:103
[16-Feb-2021 15:49:39 UTC] PHP  15. trigger_error() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:5311
```
I added these commits to resolve the issue.
https://github.com/impress-org/givewp/compare/5921f4438791...ff8264f4c1a9

Note: I left a todo in `LogoutRoute::permissionsCheck`.

ravinderk reviewed (changes requested):
@henryholtgeerts please resolve issues that are reported in slack.

henryholtgeerts commented:
@ravinderk The discussed changes have been made! See these commits for more information: https://github.com/impress-org/givewp/pull/5569/commits/281e8214c7381bc79f034dc9474975bb03f7c8d4, https://github.com/impress-org/givewp/pull/5569/commits/21331e3649ba002342243082fa20789b092a8765

ravinderk reviewed (changes requested):
@henryholtgeerts We are almost there. We store donor id in donation meta instead of the post author. Can you update the SQL query because currently, i do not see correct data on the donor profile page?
![image](https://user-images.githubusercontent.com/1784821/108101291-0924ec80-70ad-11eb-8acd-beda97c1ee4c.png)

henryholtgeerts commented:
@ravinderk Please complete another review of this PR, all changs you suggested have been implemented.

JasonTheAdams commented on src/DonorProfiles/Repositories/Donations.php:
```diff
@@ -21,10 +24,11 @@ public function getDonationCount( $donorId ) {
-				ON revenue.donation_id = posts.ID
-				WHERE posts.post_author = %d
-				AND posts.post_status IN ( 'publish', 'give_subscription' )
+				INNER JOIN {$wpdb->posts} as posts ON revenue.donation_id = posts.ID
+				INNER JOIN {$wpdb->prefix}give_donationmeta as donationmeta ON revenue.donation_id = donationmeta.donation_id
+				WHERE donationmeta.meta_key = '_give_payment_donor_id'
+					AND donationmeta.meta_value = %d
+					AND posts.post_status IN ( 'publish', 'give_subscription' )
```
All of these queries are identical aside from the SELECT. I suggest making a private `Donations::getDonationAggregate` method:
```php
private function getDonationAggregate($rawAggregate) {
	return DB::get_row(
		DB::prepare("
			SELECT {$rawAggregate}
			FROM {$wpdb->give_revenue} as revenue
				INNER JOIN {$wpdb->posts} as posts ON revenue.donation_id = posts.ID
				INNER JOIN {$wpdb->prefix}give_donationmeta as donationmeta ON revenue.donation_id = donationmeta.donation_id
			WHERE donationmeta.meta_key = '_give_payment_donor_id'
				AND donationmeta.meta_value = %d
				AND posts.post_status IN ( 'publish', 'give_subscription' )
		");
	);
}
```

Used as:
```php
return $this->getDonationAggregate('count(revenue.id)');
```

henryholtgeerts replied:
Good thinking, and glad to get the DB class into play here. Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/b00923a6246af75dd9c941d5d3c12fd622012245

JasonTheAdams commented on src/DonorProfiles/resources/js/app/components/tab-menu/utils/index.js:
```diff
@@ -0,0 +1,11 @@
+
+export const logoutWithAPI = () => {
+	return axios.post( getAPIRoot() + 'give-api/v2/donor-profile/logout', {
+		headers: {
+			'X-WP-Nonce': getAPINonce(),
+		},
```
If you have something consistent like this that needs to be done with _every_ axios request, then it's handy to use the [axios.create](https://github.com/axios/axios#axioscreateconfig), which then gets used by all of the places you're otherwise using axios.

henryholtgeerts replied:
This is a good idea -- I'm gonna hold off on implementing for a future code polish PR, since I'd rather have consistent implementation across the domain (which we currently have) over introducing this syntactical sugar here only.

JasonTheAdams reviewed (changes requested):
Great work, @henryholtgeerts! A couple _very quick_ code tweaks and this is good to go!

henryholtgeerts commented on src/DonorProfiles/Routes/LogoutRoute.php:
```diff
@@ -0,0 +1,89 @@
+
+	/**
+	 * Check permissions
+	 * @todo: need to check if donor logged as wp user or email access
```
Ah, yes! Shoulda been cleared up. Done, ref: https://github.com/impress-org/givewp/pull/5569/commits/d0952f5f34dd788caa4c20c62436c19b077f8874

JasonTheAdams reviewed (approved):
Great work, @henryholtgeerts! 🚀

ravinderk commented on src/DonorProfiles/Repositories/Donations.php:
```diff
@@ -46,27 +31,9 @@ public function getDonationCount( $donorId ) {
-
-		return Money::ofMinor( $result->amount, give_get_option( 'currency' ) )->getAmount();
+		$aggregate = $this->getDonationAggregate( 'sum(revenue.amount)', $donorId );
+		error_log( serialize( $aggregate ) );
```
@henryholtgeerts Can we remove this code?

ravinderk commented on src/DonorProfiles/Routes/LoginRoute.php:
```diff
@@ -0,0 +1,102 @@
+						'status'        => 400,
+						'response'      => 'incorrect_password',
+						'body_response' => [
+							'message' => __( 'The provided password was incorrect.', 'give' ),
```
@henryholtgeerts Can we use `esc_html__`

ravinderk commented on src/DonorProfiles/Routes/LoginRoute.php:
```diff
@@ -0,0 +1,102 @@
+					'status'        => 400,
+					'response'      => 'unidentified_login',
+					'body_response' => [
+						'message' => sprintf( __( 'A record for the provided login (%s) could not be found.', 'give' ), $login ),
```
@henryholtgeerts Can we use `esc_html__`

ravinderk commented on src/DonorProfiles/Routes/LogoutRoute.php:
```diff
@@ -0,0 +1,85 @@
+				'status'        => 200,
+				'response'      => 'logout_successful',
+				'body_response' => [
+					'message' => __( 'User was logged out successfully.', 'give' ),
```
Can we use `esc_html__`?

ravinderk commented on src/DonorProfiles/ServiceProvider.php:
```diff
@@ -8,6 +8,12 @@
 use Give\DonorProfiles\App as App;
 
+use Give\DonorProfiles\Profile as Profile;
+
```
@henryholtgeerts Can we remove this extra line?

ravinderk commented on src/DonorProfiles/ServiceProvider.php:
```diff
@@ -8,6 +8,12 @@
+use Give\DonorProfiles\Routes\LoginRoute;
+use Give\DonorProfiles\Routes\LogoutRoute;
+use Give\DonorProfiles\Routes\VerifyEmailRoute;
+
```
@henryholtgeerts Can we remove this extra line?

ravinderk commented on templates/shortcode-receipt.php:
```diff
@@ -37,11 +37,11 @@
  */
-$give_receipt_args['donation_receipt']['donor'] = array(
+$give_receipt_args['donation_receipt']['donor'] = [
 	'name'    => __( 'Donor', 'give' ),
```
@henryholtgeerts Can we use `esc_html__`
